### PR TITLE
slice2py: add latest Python keywords from Python 3.11 language reference.

### PR DESCRIPTION
**Pull request by ethindp**

### CODE REVIEW

**Comment by ethindp** [comment]
Uh, I'm blind, so I can't exactly sign the agreement and email scanned copies around.

**Comment by pepone** [comment]
Hi Ethin,

Thanks for this contribution, we are sorting out the CLA signing issue. 

The update is not totally correct, we need too keep the keywords in ASCII sort order, not in alphabetical order, that is why None with capital N was at the begining; this is required because the Slice compilers do a binary search when looking up keywords.

Cheers,
Jose

**Comment by ethindp** [comment]
@pepone I know that it uses a binary search, but placing the keywords in alphabetical order instead of ASCII alphabetical order shouldn't impact the search too much. If I'm right, the computational complexity (O(k log n)) shouldn't be impacted. I think a better idea might be to update the code to use modern C++ constructs, if possible; for example, the keywords could be stored in an [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) or [`std::set`](https://en.cppreference.com/w/cpp/container/set). We could either use [`std::find`](https://en.cppreference.com/w/cpp/algorithm/find), [`std::ranges::find`](https://en.cppreference.com/w/cpp/algorithm/ranges/find), or, if we go with a set (which would make sense since that's what a keyword list is), we could use [`std::set::contains`](https://en.cppreference.com/w/cpp/container/set/contains). I do think that moving to more modern C++17/C++20 constructs might be a wise decision overall for the entie C++ mapping, not just because we might see some significant gains in terms of performance (depending on how we use the new features) but because things would most likely be significantly cleaner.

**Comment by externl** [comment]
Hi @ethindp, can you rebase on top of or merge in the latest 3.7 branch changes?

File: cpp/src/Slice/PythonUtil.cpp [hunk 1]
```diff
@@ -222,9 +222,44 @@ lookupKwd(const string& name)
...
+        "await",
+        "break",
+        "case",
+        "class",
+        "continue",
+        "def",
+        "del",
+        "elif",
+        "else",
+        "except",
```
Comment (pepone) [inline cpp/src/Slice/PythonUtil.cpp:242]: We need to add `exec` back, as it is a keyword with Python2.x which is one of the supported versions for Ice 3.7
```suggestion
        "except",
        "exec",
```

File: cpp/src/Slice/PythonUtil.cpp [hunk 2]
```diff
@@ -222,9 +222,44 @@ lookupKwd(const string& name)
...
+        "if",
+        "import",
+        "in",
+        "is",
+        "lambda",
+        "match",
+        "nonlocal",
+        "not",
+        "or",
+        "pass",
```
Comment (pepone) [inline cpp/src/Slice/PythonUtil.cpp:256]: Same for `print`, we cannot remove keywords in a patch release

```suggestion
        "pass",
        "print",
```

File: cpp/src/Slice/PythonUtil.cpp [hunk 3]
```diff
@@ -222,9 +222,44 @@ lookupKwd(const string& name)
...
+        "None",
+        "True",
+        "_",
+        "and",
+        "as",
+        "assert",
+        "async",
+        "await",
+        "break",
+        "case",
```
Comment (pepone) [inline cpp/src/Slice/PythonUtil.cpp:235]: I don't see `case` among the keywords for Python 2.x or Python 3.x, why did you add it here?

Comment (ethindp) [reply cpp/src/Slice/PythonUtil.cpp:234]: case isa keyword used in match statements. It's a soft keyword; it can be used as an identifier only outside a match statement. However, adding it seemed like a good idea since I didn't want that to possibly become a problem should slice2py's Python3 support use match statements in future.

**Comment by ethindp** [comment]
I'm unsure if my push (pull/push, rather) broke something or not. I apologize if it re-merged anything or caused other problems.

**Comment by pepone** [comment]
> I'm unsure if my push (pull/push, rather)

Everything seems good, I would merge the PR once the CI build confirms it is working.

Thanks for your contribution to Ice

**Review by pepone — APPROVED** [review]
